### PR TITLE
Implement reporting to Distribution beacon for proper download counts

### PR DESCRIPTION
**Pull request by nevalenny**
1. Report install id on Distribution setup page
2. Report release id and distribution id for private groups
3. Report release id and install id for public groups

### CODE REVIEW

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 1]
```diff
@@ -863,6 +866,24 @@ synchronized void getLatestReleaseDetails(String distributionGroupId, String upd
         } else {
             url += String.format(GET_LATEST_PRIVATE_RELEASE_PATH_FORMAT, mAppSecret, releaseHash);
         }
+
+        /* Send install id once upon in-app update if feature flag is enabled */
+        String shouldUseInstallIdFeature = mContext.getString(R.string.install_id_feature_enabled);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: This is completely harmless to do that for everyone, why insisting on having a feature flag for this?

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 2]
```diff
@@ -863,6 +866,24 @@ synchronized void getLatestReleaseDetails(String distributionGroupId, String upd
...
+        if(shouldUseInstallIdFeature != null && shouldUseInstallIdFeature.equals("True")) {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Please format file with Android Studio.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeConstants.java [hunk 3]
```diff
@@ -106,6 +106,16 @@
...
 
+    /**
+     * API parameter for install identifier.
+     */
+    static final String PARAMETER_INSTALL_ID = "install_id";
+
+    /**
+     * API parameter for last installed release hash.
+     */
+    static final String PARAMETER_LAST_INSTALLED_RELEASE_HASH = "last_installed_release_hash";
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeConstants.java]: Aren't we supposed to send release identifier instead? There is a one to many relationship between hash and identifiers: multiple release ids can have the exact same hash as its only based on version.

File: sdk/appcenter-distribute/src/debug/res/values/features.xml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="install_id_feature_enabled">True</string>
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/debug/res/values/features.xml]: Add `appcenter_distribute_` to every string resource as this gets injected into customer resources. We need proper namespacing. We do that also for texts.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 2 above)
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: `True` is a C# convention, not a Java one (`true`).
After changing to `true` you can simplify the check with a single statement which is the preferred way of checking boolean strings in java: `Boolean.parseBoolean(string)`.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java [hunk 5]
```diff
@@ -130,6 +132,16 @@ static void updateSetupUsingBrowser(Activity activity, String installUrl, String
         url += "&" + PARAMETER_REQUEST_ID + "=" + requestId;
         url += "&" + PARAMETER_PLATFORM + "=" + PARAMETER_PLATFORM_VALUE;
         url += "&" + PARAMETER_ENABLE_UPDATE_SETUP_FAILURE_REDIRECT_KEY + "=" + "true";
+
+        /* Add install_id flagged for usage with debug app only. */
+        String shouldUseInstallIdFeature = activity.getString(R.string.install_id_feature_enabled);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeUtils.java]: Reformat file with Android Studio.

File: sdk/appcenter-distribute/src/release/res/values/features.xml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+<string name="install_id_feature_enabled">False</string>
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/release/res/values/features.xml]: The released SDK is always in release by the way.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 2 above)
Comment (nevalenny) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Thanks, great tip! Fixed.

Comment (nevalenny) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Fixed

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 1 above)
Comment (nevalenny) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: This is still under construction, so we keep the whole thing under feature flags to be independent of release cycles etc. We can remove/enable it anytime. For now, I think that was a team decision, but we can reconsider this with @Meirav

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeConstants.java [hunk 7]
```diff
@@ -106,6 +111,16 @@
...
 
+    /**
+     * API parameter for install identifier.
+     */
+    static final String PARAMETER_INSTALL_ID = "install_id";
+
+    /**
+     * API parameter for last installed release hash.
+     */
+    static final String PARAMETER_LAST_DOWNLOADED_RELEASE_HASH = "downloaded_release_hash";
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeConstants.java]: Why do we report release hash? there is a one to many relationship between a single release hash and multiple release ids. Release hash is just based on packagename and version, you can have multiple releases with same hash very easily.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java [hunk 8]
```diff
@@ -445,6 +467,24 @@ public void happyPathUntilHangingCallWithToken() throws Exception {
         when(UUIDUtils.randomUUID()).thenReturn(requestId);
         when(PreferencesStorage.getString(PREFERENCE_KEY_REQUEST_ID)).thenReturn(requestId.toString());
 
+        /* Mock install id from AppCenter */
+        final UUID installId = UUID.randomUUID();
+        when(AppCenter.getInstallId()).thenReturn(new AppCenterFuture<UUID>() {
+            @Override
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: add blank line, prefer using mocking framework than using anonymous inner classes. See all the test code for many examples.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java [hunk 9]
```diff
@@ -564,6 +605,24 @@ public void happyPathUntilHangingCallWithoutToken() throws Exception {
...
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: same comment

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java [hunk 10]
```diff
@@ -683,6 +743,24 @@ public void setUrls() throws Exception {
...
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: same comment

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java [hunk 11]
```diff
@@ -718,6 +797,24 @@ public void computeHashFailsWhenOpeningBrowser() throws Exception {
         /* Mock package manager. */
         when(mPackageManager.getPackageInfo("com.contoso", 0)).thenThrow(new PackageManager.NameNotFoundException());
 
+        /* Mock install id from AppCenter */
+        final UUID installId = UUID.randomUUID();
+        when(AppCenter.getInstallId()).thenReturn(new AppCenterFuture<UUID>() {
+            @Override
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: same comment

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java [hunk 12]
```diff
@@ -738,6 +835,25 @@ public void disableBeforeStoreToken() {
         /* Start and resume: open browser. */
         UUID requestId = UUID.randomUUID();
         when(UUIDUtils.randomUUID()).thenReturn(requestId);
+
+        /* Mock install id from AppCenter */
+        final UUID installId = UUID.randomUUID();
+        when(AppCenter.getInstallId()).thenReturn(new AppCenterFuture<UUID>() {
+            @Override
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: same comment

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeConstants.java (hunk 7 above)
Comment (nevalenny) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/DistributeConstants.java]: Will remove it, the description is legacy and misleading.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java (hunk 8 above)
Comment (nevalenny) [reply sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeApiSuccessTest.java]: Fixed that, it was looking too ugly, agree.

**Comment by guperrot** [comment]
Restarting build by close/reopen as there was a transient issue on the build machine.

**Comment by nevalenny** [comment]
@guperrot yes, I saw them started failing. Will continue working on those two today.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/CheckDownloadTask.java [hunk 13]
```diff
@@ -154,6 +157,15 @@ protected DownloadProgress doInBackground(Void... params) {
                     } else {
                         distribute.completeWorkflow(mReleaseDetails);
                     }
+
+                    /* Add downloaded release hash to report after installation */
```
Comment (MatkovIvan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/CheckDownloadTask.java]: Dots at the end of the comments. Same in some other places.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/AbstractDistributeTest.java [hunk 14]
```diff
@@ -211,16 +211,16 @@ public String answer(InvocationOnMock invocation) throws Throwable {
 
             @Override
             public Void answer(InvocationOnMock invocation) throws Throwable {
-            when(mDialog.isShowing()).thenReturn(true);
-            return null;
+                when(mDialog.isShowing()).thenReturn(true);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/AbstractDistributeTest.java]: is that tabs? we use only spaces to indent, not sure if github is tricking me.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 15]
```diff
@@ -1009,6 +1025,39 @@ private synchronized void handleApiCallSuccess(Object releaseCallId, String rawR
...
+    /**
+     * Get reporting parameters for new release installation.
+     *
+     * @param isPublic            are the parameters for public group or not.
+     *                            For public group we report install_id and release_id.
+     *                            For private group we report distribution_group_id and release_id.
+     * @param distributionGroupId distribution group id.
+     */
+    @NonNull
+    private String getReportingParametersForNewInstallation(boolean isPublic, String distributionGroupId) {
```
[the remark this answers is not in the document]
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Renamed to `getReportingParametersForUpdatedRelease`. Does this look good to you?

(Yury is OOF so I'll take some of his tasks)

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 16]
```diff
@@ -1009,6 +1025,39 @@ private synchronized void handleApiCallSuccess(Object releaseCallId, String rawR
...
+        if (!TextUtils.isEmpty(lastDownloadedReleaseHash)) {
+            String currentInstalledReleaseHash = computeReleaseHash(mPackageInfo);
+            if (lastDownloadedReleaseHash.equals(currentInstalledReleaseHash)) {
+                AppCenterLog.debug(LOG_TAG, "Current release was updated but not reported yet, reporting..");
+                if (isPublic) {
+                    reportingParameters += "&" + PARAMETER_INSTALL_ID + "=" + AppCenter.getInstallId().get();
+                } else {
+                    reportingParameters += "&" + PARAMETER_DISTRIBUTION_GROUP_ID + "=" + distributionGroupId;
+                }
+                Integer lastDownloadedReleaseId = PreferencesStorage.getInt(PREFERENCE_KEY_DOWNLOADED_RELEASE_ID);
```
[the remark this answers is not in the document]
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Fixed. Thank you.

(Yury is OOF so I'll take some of his tasks)

**Comment by guperrot** [comment]
@roland-chernov-akvelon sounds good. By the way on previous PR that I merged there was some warnings like unused imports or variables to be inlined. Please use analyze code to find those, some are existing warnings but if you can fix the one related to DistributeInfoTracker in a separate PR that would be great.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/AbstractDistributeTest.java (hunk 14 above)
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/AbstractDistributeTest.java]: No, it's spaces. GitHub diffs seem confusing a little bit.

Please see the screenshot:
Screenshot

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeDownloadTest.java [hunk 17]
```diff
@@ -59,7 +66,10 @@
 import static org.powermock.api.mockito.PowerMockito.verifyStatic;
 import static org.powermock.api.mockito.PowerMockito.whenNew;
 
+@PrepareForTest(DistributeUtils.class)
 public class DistributeBeforeDownloadTest extends AbstractDistributeTest {
+    @Mock
```
[the remark this answers is not in the document]
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeBeforeDownloadTest.java]: Fixed. Thank you.

**Comment by roland-chernov-akvelon** [comment]
@guperrot Sure, will do.

**Comment by roland-chernov-akvelon** [comment]
@guperrot PR to fix warnings was created. Please take a look #616.

**Comment by guperrot** [comment]
I tested end 2 end with private and public, merging now.

**Pull request merged**